tinygrad/engine/schedule.py

```diff
@@ -193,17 +205,23 @@ def create_schedule(outs:List[LazyBuffer], seen:Optional[Set[LazyBuffer]]=None)
         in_degree[assign_targets[x]] += 1
       if x in prescheduled: in_degree[out] += 1
 
-  queue = deque(out for out in prescheduled if in_degree[out] == 0)
-  schedule: List[ScheduleItem] = []
+  queue: Deque[Tuple[int,LazyBuffer]] = deque((0,out) for out in prescheduled if in_degree[out] == 0)
+  groups: DefaultDict[Tuple,List[ScheduleItem]] = defaultdict(list)
   while queue:
-    buf = queue.popleft()
+    level, buf = queue.popleft()
```

Comment:
I don't understand what level does here

Reply:
I'm borrowing the idea from level order traversal of binary trees. The way we can think about "level" in ScheduleItems toposort is that the items in the same level can run in parallel.

Reply:
Hmm, I think a lot more fusion might be possible.

I don't love the `_fuse_group` function. Have you tried doing this instead by relaxing scheduler constraints on line 157? Ideally this should be line neutral.

Reply:
I think 157 only checks the case for reduces.
Adam's weight update realizes happen in `_recurse_lb`, we need to graph the realizes.